app/ide-desktop/lib/dashboard/src/services/LocalBackend.ts

```diff
@@ -21,6 +90,66 @@ function ipWithSocketToAddress(ipWithSocket: projectManager.IpWithSocket) {
   return backend.Address(`ws://${ipWithSocket.host}:${ipWithSocket.port}`)
 }
 
+// ======================================
+// === Functions for manipulating ids ===
+// ======================================
+
+/** Create a {@link backend.DirectoryId} from a path. */
+function newDirectoryId(path: string) {
+  return backend.DirectoryId(`${backend.AssetType.directory}-${path}`)
+}
+
+/** Create a {@link backend.ProjectId} from a path. */
+function newProjectId(path: string) {
+  return backend.ProjectId(`${backend.AssetType.project}-${path}`)
+}
+
+/** Create a {@link backend.FileId} from a path. */
+function newFileId(path: string) {
+  return backend.FileId(`${backend.AssetType.file}-${path}`)
+}
+
+/** The subset of {@link backend.AssetId} that is handled by a {@link LocalBackend}. */
+type LocalAssetId = backend.DirectoryId | backend.FileId | backend.ProjectId
+
+/** A structure with an `id`. */
+interface HasId<Id extends backend.AssetId> {
+  readonly id: Id
+}
+
+// `A extends A` is required to trigger distributive conditional types:
+// https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types
+/** An asset type and its corresponding id. */
+type AssetTypeAndId =
+  Extract<backend.AnyAsset, HasId<LocalAssetId>> extends infer A
+    ? A extends A
+      ? Pick<A, keyof A & ('id' | 'type')>
+      : never
+    : never
+
+export function extractTypeAndId<Id extends backend.AssetId>(
+  id: Id
+): Extract<AssetTypeAndId, HasId<Id>>
+/** Extracts the asset type and its corresponding internal ID from a {@link backend.AssetId}.
+ * @throws {Error} if the id has an unknown type. */
+export function extractTypeAndId<Id extends backend.AssetId>(id: Id): AssetTypeAndId {
+  const [, typeRaw, idRaw] = id.match(/(.+?)-(.+)/) ?? []
+  switch (typeRaw) {
+    case backend.AssetType.directory: {
+      return { type: backend.AssetType.directory, id: backend.DirectoryId(idRaw ?? '') }
```

Comment:
How come `id` can be an empty string?

Reply:
it shouldn't be able to, but we cannot statically assert that even if we wanted to, because TS doesn't parse regexes and so we don't know the number of matches at the type level